README.md

```diff
@@ -1,277 +1,899 @@
+batcher based composition to *not be trustless* (without the composition logic, more dApps can fit
+in a transaction).
+- No delegation control or voting control while using the DEX - DEXs try to issue governance tokens
+to get around this but fair distribution of the governance tokens is rarely accomplished (it is a
+very hard problem). Even if the governance tokens were distributed fairly, they are largely just for
+show since the DEX creators can still choose to go against the vote. There is no trustless
+connection between the governance tokens and what actually happens with the stake/dApp. Finally,
+even if there was a trustless connection, there is no way to actually have all users express their
+own stake preferences when the stake is decided democratically. It is almost inevitable that the
+voting outcome will go against the wishes of some users. *This is in direct violation of the
+security assumptions of PoS.* No matter how you look at it, LP based dApps are an existenstial
+problem for PoS blockchains. As LP based dApps become more adopted, the underlying PoS blockchain
+will only become more centralized. LPs can never be used as the foundation for a trustless and
+decentralized PoS blockchain economy.
+
+DEXs that hope to become the foundation for a healthy and fully trustless, decentralized economy
+must adhere to a radically different approach that takes full advantage of the composability and
+availability guarantees offered by the eUTxO model while not undermining the security of the very
+blockchain the economy is built on.
+
+## The Cardano-Swaps Protocol
+
+Cardano-Swaps is an order-book based DEX that takes inspiration from Axo's programmable swaps
+[design](https://www.axo.trade/whitepaper.pdf), adds delegation control as a foundational feature,
+naturally incentives its own liquidity, and, through the use of Beacon Tokens, removes the need for
+specialized indexers. In short, it solves all of the pressing issues that made order-book DEXs
+inferior to LP based DEXs. 
+
+### High-Level Overview
+
+The protocol is comprised of two types of composable swaps: a one-way swap and a two-way swap.
+One-way swaps are effectively just limit orders (ie, they can only be executed at the specified
+price or better). Most users will likely use the one-way swap.
+
+Two-way swaps are a mechanism for profitably providing liquidity to the protocol without having to
+resort to yield farming. These two-way swaps can go in either direction (eg, a two-way ADA <-> ERGO
+swap can go ADA -> ERGO or ERGO -> ADA). By allowing the liquidity provider to specify separate
+prices for each direction, they can profit from either direction of the swap. For example, if the
+market rate for DJED <-> USDC is 1:1, Alice can set DJED -> USDC to 1.01 USDC per DJED and USDC ->
+DJED to 1.01 DJED per USDC. This means Alice makes a 1% return no matter what direction is used.
```

Comment:
>  In other words, a permanent price change would not protect the two-way swap user from losses?

I'm not sure I follow. No loss has occurred in your example (assuming the exchange rate is 1:1): before the swap the consolidation, the total value is $200; after the consolidation, the total value is $201. That is a profit. The fact that the price of DJED in USDC remained above 1 is not an issue, imo. 

Two-way swaps are meant to be a "last resort" swap because you will always have to pay a slight premium, no matter the swap direction. But that is not the same as saying there is no economic utility for this feature. If Alice has DJED but really needs USDC, she will use whatever she can to get the USDC. I think most users will be okay paying the slight premiums in this scenario. Since two-way swap providers are competing with each other, the premiums will approach the true value. Basically, if Alice really needs USDC, she has a choice of missing the opportunity that requires USDC or paying up marginally. 

Consider this, with all the stablecoins being created, liquidity is fractured into "silos" (one for each stablecoin). Two-ways swaps makes the liquidity of stablecoins freely convertible at a slight premium. I think there is tremendous utility for this feature, especially since the conversion can happen in the same transaction as other p2p protocol executions (like buying an options contract that requires DJED). Personally, if I have DJED but need USDC to buy an options contract, I would be perfectly fine paying a 1% premium to acquire the USDC (assuming the options contract can recoup the expense).

> I do like the idea of this two-way swap for things like stable coins, where the price is expected to fluctuate around parity...

This is the number one use case for two-way swaps. They should be seen as active trades, not "set and forget" trades. You are correct that they are more suited for non-volatile pairs but there could still be value in active traders using them for volatile pairs, although this then carries a higher risk of loss.

The way I see it, two-way swaps incentivize providing liquidity to main trading pairs such as stablecoin <-> stablecoin or ADA <-> stablecoin. Then, the arbitragers take this liquidity and spread it. There do not need to be two-way swaps for all trading pairs for their to be liquidity on the DEX. Only the entry and exit points (the main trading pairs) need liquidity.

Reply:
> I'm not sure I follow. No loss has occurred in your example (assuming the exchange rate is 1:1):

That is because we are assuming the stable coins will remain at 1:1 exchange rate.  If one stable coin permanently de-pegs then the owner of the two-way swap will end up holding the de-pegged coin only and will therefore suffer loss if the price never re-peggs.  Similarly if you create a two-way swap for any other asset pair, then you will end up holding only the asset which becomes relatively less valuable.  For example, if you create an ADA <-> DJED two-way swap at 1 ADA -> 1.01 DJED and 1 DJED -> 1.01 Ada, then if Ada goes above 1.01 DJED then your swap contract will now only hold DJED because arbitragers will fully execute it in the DJED -> ADA direction.  In other words as the price of Ada increases you will be left holding only the less valuable asset.

The way I see the two-way swap is a bit like an individual AMM swap but instead of having an exchange rate that varies in an attempt to offset the relative price change (eg. using constant product formula), instead there is like a two-way price cliff where once the price is exceeded in one direction then the contract will get fully executed in the direction of the weaker asset.

I like the concept for stable coins where they are expected to fluctuate around parity with each other, but I can't see it working well for any other asset pairs.  I don't think it is fundamentally unsound or anything.  I just wouldn't use a two-way swap myself for anything other than a stable coin pair and only if I was extremely confident in each coin ultimately returning to peg.

For stable coins, I do agree this is likely to improve liquidity because I believe people will want to create such swaps in order to accumulate stable coins as the price fluctuates either side of parity.

Reply:
> ... there is like a two-way price cliff where once the price is exceeded in one direction then the contract will get fully executed in the direction of the weaker asset.

This is why it is still an active trade. Stablecoins are the safest use-case (as you said, a de-peg is still a risk) but I do think some users would make two-way swaps for larger assets <-> stablecoin, like ADA <-> DJED. I'd personally be interested in it during sideways market movements and manage my risk by limiting the amount of assets in the ADA <-> DJED swap and charging a very high premium in the direction I think is weakest. I would expect the vast majority of two-way swaps are for stablecoin <-> stablecoin but I do think there will be at least *some* two-way swaps for other pairs. Professional traders can charge higher premiums for these since they have less competition. I think it is an open question as to whether there will be *enough* two-way swaps for the other pairs but I think there is plenty of incentive for it. Arbitragers themselves may be willing to pay premiums as high as 30% for a swap like ADA <-> DJED if the arbitrage opportunities are profitable enough. Considering the very nature of illiquidity, I think there actually will be enough profit potential for arbitragers to accept the 30% premium. 

I will concede that this is speculation on my part, though. I am personally interested in building an automated arbitrager and imagining what I would do.

Reply:
I can see how you could use the two-way swaps as part of an automated arbitrager with it updating the exchange rates as the price changes.

Reply:
Thinking about this yesterday made me realize there is a hole in the incentives that I need to patch before I release the beta. This hole mainly impacts swaps like ADA <-> stablecoin. I don't think it really impacts stablecoin <-> stablecoin. Basically, the chaining of swaps opens up the possibility of a flash swap (name deliberately similar to flash loan) where a swap's assets are taken and returned in the same transaction. Since the swap's balance didn't change, the protocol will not charge the premium. I'm going to explain it here, partly to make sure I understand it fully myself and partly to see if anyone has feedback on it.

The way I was initially imagining arbitrage was like this: `ADA -> x -> y -> z -> ADA`. Each arrow is a swap. In this scenario, the start and end swaps are different swaps. But it is actually possible to do this: `ADA -> x -> y -> z -> x -> ADA`. In this scenario, the start and end swaps can be different swaps but don't have to be. When `z -> x` is executed, the resulting `x` asset can be returned to the `ADA -> x` swap and the ada used in that swap can be reclaimed. So on net, the `ADA -> x` swap had no change in the balance of the swap UTxO. This swap is a flash swap: `x` is taken and returned in the same transaction. Since the balance of the swap did not change, the swap does not require any payment. I have manually confirmed that this is indeed possible.

This technically does not hurt anyone since no assets were officially taken without paying. The issue is the arbitrager used the flash swap to make a profit but did not pay the user for providing the liquidity. Since no premium needs to be paid to do this, there is a massive incentive for arbitragers to use flash swaps. Since stablecoin <-> stablecoin swaps are unlikely to be chained, I don't think this really impacts them. This definitely impacts all other swaps, though. While I think there are more use cases for providing liquidity for other trading pairs than just arbitrage, providing liquidity for arbitragers is the main incentive since they will likely be the most active. So the arbitrager being able to use the swap's liquidity without paying severely dampens the incentive for users to provide liquidity.

Considering what smart contracts can see about the transaction and given the redundant executions, there is no cheap way to tell how much of the flash swap's assets were used and charge the premium accordingly. But this actually may not be an issue. The premium is meant for scenarios where the swap's assets were actually converted. Since no assets were actually taken, I do not think the premium needs to be charged. Instead, a flat fee for using the swap seems more appropriate.

At a high level, if I added a `minimumGive` field to the datums, users can specify how much they want of the ask asset per execution. This would apply both to normal swaps and flash swaps. For a normal swap, if Alice has 10 ada in a `ADA -> DJED` swap and specifies a `minimumGive` of 2 djed at a 1:1 ratio, if Bob wants to swap with Alice, he must give at least 2 djed in the swap. It is still possible for Bob to take any amount of ada up to the ratio, but taking less than 2 ada would be a loss for Bob. Bob is incentivized *not* to do this.

For flash swaps, it does not matter that none of the swap's offer asset was taken; the `minimumGive` will still be  enforced. In this scenario, this `minimumGive` is the fee the liquidity provider is charging for allowing the arbitrager to use their swap as a flash swap. If the swap is actually converted, the premium will still be enforced. Considering arbitragers will prefer this fixed fee over the premium, I think this actually makes it safer for users to provide liquidity of larger trading pairs - most swaps will likely be arbitragers using the swap as a flash swap.

I had considered a `minimumTake` field instead but the quantity of the offer asset is what is important to arbitragers. If the quantity must decrease which each execution, eventually the swap would be unusable by the arbitragers. The `minimumGive` allows the swap to be used indefinitely while still having the same result.

Reply:
I've been testing this more and I may be wrong about the flash swap being useful here. Intuitively, since the flash swap doesn't change, it shouldn't be needed. And it turns out, it is not needed. The above example of `ADA -> x -> y -> z -> x -> ADA` can actually happen without the `ADA -> x` swap. It would effectively be this:

```mermaid
graph TD;
    X-->Y;
    Y-->Z;
    Z-->X;
```
The `x` that is needed to unlock the `y` in the first swap comes from the last swap. It is circular. So the inclusion of the `ADA -> x` swap does nothing but increase the transaction fee. There may actually be no need for the `minimumGive` but I need to think about this more.

Reply:
In light of this, it does sound like the protocol parameters from which the minimum Tx fees are derived, having been chosen primarily to prevent DDoS on the network, would also cover this similar case.

Reply:
For what it is worth, I agree with your reasoning here and agree that the extra complexity is not needed, and would make your design less neat.

Reply:
I can't come up with any situation where flash swaps would be useful. Any time you could use a flash swap, the circular swap would be cheaper. So I don't see a need to add the `minimumGive`.